Add technique to reduce Rayleigh contribution at high zenith angles

simonrp84 opened the pull request:
At present, the Rayleigh correction code calculates the Rayleigh reflectance as a function of zenith and azimuth angles. This works well and is accurate in many cases, but at high zenith the assumptions used in the Look-Up Tables behind the calculation break down and can produce unphsyically high Rayleigh contributions.

In this PR, I add a simple (non-radiative transfer) method for reducing this effect, which results in more visually pleasing images at high zenith angles. The method linearly reduces the Rayleigh contribution between a threshold angle and whatever the maximum zenith angle for the LUTs is: `new_rayleigh = old_rayleigh * factor(zenith, thresh_zen, maxzen)`. The relative strength of this effect can be altered via a new variable in the `get_reflectance` function call, `reduce_strength`. This correction is only applied to the solar zenith angle, not to high satellite zeniths.

By default, this new technique is *disabled*, and is only enabled if the user passes `reduce_hizen=some_float` to the Rayleigh code's `get_reflectance` function.
Attached is an example of these updates applied to Him-8/AHI data.

The original `true_color_reproduction` composite produced using `main` branches of satpy and pyspectral:
![orig_rez](https://user-images.githubusercontent.com/13449576/146533251-a31caa5c-ca75-4fc4-b7eb-980086141e9b.jpg)
The same composite with this PR enabled, `thresh_zen` is set to 75 degrees and `reduce_strength` is 0.75.
![enabled_75deg_0 7str_rez](https://user-images.githubusercontent.com/13449576/146533253-c550298c-5d05-4356-8453-3955352df1ef.jpg)

Note: As part of this PR I have also removed a bunch of imports from the tests that were only suitable for python 2.x, and were thus unused.

 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [x] Tests passed: Passes ``pytest pyspectral`` <!-- for all non-documentation changes) -->
 - [ ] Passes ``flake8 pyspectral`` <!-- remove if you did not edit any Python files -->
 - [ ] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

adybbroe commented:
@simonrp84 Could you show the raw True Color image as well.

simonrp84 commented:
@adybbroe Of course! Here's the same himawari timeslot with the `true_color` composite.
Without PR:
![old](https://user-images.githubusercontent.com/13449576/146539617-83c2b235-fb42-4a15-afed-b3e76edf37a6.jpg)
With PR:
![new](https://user-images.githubusercontent.com/13449576/146539634-d65eeb8b-a4e1-4b6b-becd-3556309b9d28.jpg)
I haven't fiddled with the settings for this one, so it's not optimised, but the high-SZA part over Australia is no longer overly-red, so IMO the PR works nicely for this composite too.

adybbroe commented:
@simonrp84 I was thinking the raw (uncorrected) true color. I am a bit confused now, isn't the "Without PR" the same as the `true_color_reproduction`?
And I think you are right, but I am just disturbed by the blue (rayleigh scattering) rim close to the terminator. But I suppose that is the prize you have to pay?
@rolle-at-work Any comments?

simonrp84 commented:
Aah sorry @adybbroe. Here's the `true_color` image with no corrections at all (no rayleigh, no sun zenith):
![nocorr](https://user-images.githubusercontent.com/13449576/146545710-245a88f9-cbd4-4676-897e-3077c4502a8a.jpg)
and here it is with sun zenith correction but no Rayleigh correction:
![noray](https://user-images.githubusercontent.com/13449576/146545737-39150a99-ebd8-407b-8293-6608f545d105.jpg)
Both of the above were made without my PR.

>"I am a bit confused now, isn't the "Without PR" the same as the true_color_reproduction?"

If you load the "without PR" image from above in a separate browser tab you can compare directly to the `true_color_reproduction`. They're similar, but not the same.

> And I think you are right, but I am just disturbed by the blue (rayleigh scattering) rim close to the terminator. But I suppose that is the prize you have to pay?

Agree that there's too much blue, but that's because I didn't tweak the `reduce_strength`. Below is an example where I've reduced the strength from `1.0` (default) down to `0.3`. This reduces the blue a lot. It seems, as you might expect, that the stength required varies with the composite.
![03ray](https://user-images.githubusercontent.com/13449576/146546547-0c5650f0-a78c-4531-abdd-a80c884bde21.jpg)

adybbroe commented:
Nice, thanks @simonrp84 !

mraspaud commented on pyspectral/tests/test_rayleigh.py:
```diff
@@ -201,6 +196,18 @@ def test_get_reflectance(self, get_effective_wvl,
             sun_zenith, sat_zenith, azidiff, 'M2', blueband)
         self.assertTrue(np.allclose(retv, TEST_RAYLEIGH_RESULT2))
 
+        # Test the Rayleigh reduction code
```
Could this be a separate test function/method?

mraspaud commented on pyspectral/tests/test_rayleigh.py:
```diff
@@ -242,6 +249,20 @@ def test_get_reflectance_dask(self, get_effective_wvl,
         self.assertTrue(np.allclose(retv, TEST_RAYLEIGH_RESULT2))
         self.assertIsInstance(retv, da.Array)
 
+        # Test the Rayleigh reduction code
```
Could this be a separate test function/method?

mraspaud commented on pyspectral/rayleigh.py:
```diff
@@ -237,7 +239,20 @@ def _do_interp(minterp, sunzsec, azidiff, satzsec):
 
-    def get_reflectance(self, sun_zenith, sat_zenith, azidiff, bandname, redband=None):
+    @staticmethod
+    def _do_reduce_effect(zenith, rayref, thresh_zen, maxzen, strength):
```
Could we have a more specific name, like eg: `_reduce_rayleigh_correction_at_high_zenith_angles`?

adybbroe commented on pyspectral/rayleigh.py:
```diff
@@ -200,15 +202,15 @@ def get_effective_wavelength(self, bandname):
+                        'det-1']['wavelength'], rsr.rsr[bandname]['det-1']['response']
 
-        cwvl = get_central_wave(wvl, resp, weight=1. / wvl**4)
+        cwvl = get_central_wave(wvl, resp, weight=1. / wvl ** 4)
```
What is happening here? My flake8/pep8 is fine with the `wvl**4` - isn't that how it supposed to be?
https://realpython.com/python-pep8/

mraspaud commented on pyspectral/rayleigh.py:
```diff
@@ -237,7 +239,20 @@ def _do_interp(minterp, sunzsec, azidiff, satzsec):
+        return rayref * factor
+
+    def get_reflectance(self, sun_zenith, sat_zenith, azidiff, bandname, redband=None,
+                        reduce_hizen=None, reduce_strength=1.):
```
Could we combine `reduce_hizen` and `reduce_strength` into one parameter? eg if we kept only reduce_strength, setting it to 0 could be equivalent to skipping this?

Also some explanation in the docstring on what the new argument(s) mean would be nice.

adybbroe commented on pyspectral/rayleigh.py:
```diff
@@ -228,16 +230,29 @@ def get_reflectance_lut(self):
-        return self._rayl, self._wvl_coord, self._azid_coord,\
-            self._satz_sec_coord, self._sunz_sec_coord
+        return self._rayl, self._wvl_coord, self._azid_coord, \
+               self._satz_sec_coord, self._sunz_sec_coord
```
Please leave as it was. Is it some kind of auto correct you have?

pnuu replied:
```suggestion
        return (self._rayl, self._wvl_coord, self._azid_coord,
                self._satz_sec_coord, self._sunz_sec_coord)
```

adybbroe replied:
Yes, thanks, choose that @simonrp84

mraspaud reviewed (commented):
Thanks for submitting this PR! I like that you took the time to clean up old stuff and rename some abiguous variable names :)

Some minor comments, the principles look good to me.

adybbroe commented on pyspectral/tests/test_rayleigh.py:
```diff
@@ -201,6 +203,18 @@ def test_get_reflectance(self, get_effective_wvl,
+        retv = self.viirs_rayleigh.get_reflectance(
+            sun_zenith, sat_zenith, azidiff, 'M2', blueband, reduce_hizen=60, reduce_strength=1.5)
+        self.assertTrue(np.allclose(retv, TEST_RAYLEIGH_RESULT_R2))
+
```
Could you put the reduction part in a separate test function? The test functions are already too long. Preferably a test-function should test only one thing.

adybbroe commented on pyspectral/tests/test_rayleigh.py:
```diff
@@ -242,6 +256,20 @@ def test_get_reflectance_dask(self, get_effective_wvl,
+        retv = self.viirs_rayleigh.get_reflectance(
+            sun_zenith, sat_zenith, azidiff, 'M2', blueband, reduce_hizen=60, reduce_strength=1.5)
+        self.assertTrue(np.allclose(retv, TEST_RAYLEIGH_RESULT_R2))
+        self.assertIsInstance(retv, da.Array)
```
Same again, test preferably only one thing per test-function.

adybbroe commented on pyspectral/rayleigh.py:
```diff
@@ -228,16 +230,29 @@ def get_reflectance_lut(self):
+        factor = clip(factor, 0, 1)
+        return rayref * factor
+
+    def get_reflectance(self, sun_zenith, sat_zenith, azidiff, bandname, redband=None,
```
Nice work @simonrp84 But the `get_reflectance` method is getting increasingly complex and have too many input arguments. But that was already bad before. Could the effect reduction not be applied after, as a separate function on the standard correction?

mherbertson commented:
I had a quick look with one timeframe, the results are great with plenty of room to fine tune depending on composite.

`true_color_reproduction` as currently:
![202112170850_TCR](https://user-images.githubusercontent.com/57386258/146633993-8cab61d5-17e4-480e-96ff-ddab3d1fcb8f.jpg)
JMA True Color Reproduction:
![202112170850_TCR_JMA_FDK](https://user-images.githubusercontent.com/57386258/146634019-de5a691e-371f-4cf3-b95b-a98b0b2a7493.jpeg)
`true_color_reproduction` with threshold 70. and strength 0.5:
![202112170850_TCR_70_05](https://user-images.githubusercontent.com/57386258/146634066-a0539351-0582-47a4-9353-aafbd567cde6.jpg)
`true_color_reproduction` with night ir:
![202112170850_TCR-IR_70_05](https://user-images.githubusercontent.com/57386258/146634079-96160a84-66eb-4036-ba4d-b88af4e3c540.jpg)

Nice work @simonrp84

rolle-at-work commented:
I'm already on vacation 🙂
If still in need, i will check when I'm back.

All the best
Rolle


-------- Original message --------
From: Adam Dybbroe ***@***.***>
Date: 17/12/2021 13:04 (GMT+01:00)
To: pytroll/pyspectral ***@***.***>
Cc: Scheirer Ronald ***@***.***>, Mention ***@***.***>
Subject: Re: [pytroll/pyspectral] Add technique to reduce Rayleigh contribution at high zenith angles (PR #141)


@simonrp84<https://github.com/simonrp84> I was thinking the raw (uncorrected) true color. I am a bit confused now, isn't the "Without PR" the same as the true_color_reproduction?
And I think you are right, but I am just disturbed by the blue (rayleigh scattering) rim close to the terminator. But I suppose that is the prize you have to pay?
@rolle-at-work<https://github.com/rolle-at-work> Any comments?

—
Reply to this email directly, view it on GitHub<https://github.com/pytroll/pyspectral/pull/141#issuecomment-996671371>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANYT5KF32WRVMV4LHRJPDGTURMRNDANCNFSM5KIRJYPA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

adybbroe commented:
@simonrp84 We need a small usage example for the documentation. I can help you if you like?

simonrp84 commented:
Thanks @adybbroe - typical research coder, I did all the fun programming stuff and forgot the docs! I have updated the documentation now and corrected a few typos along the way.

mraspaud commented on doc/rayleigh_correction.rst:
```diff
@@ -110,9 +109,32 @@ if you want another setup, e.g.:
+  >>> sunz = np.array([[32., 40.], [80., 88.]])
+  >>> satz = np.array([[45., 20.], [46., 21.]])
+  >>> ssadiff = np.array([[110, 170], [120, 180]])
+  >>> refl_cor_m2 = ufo.get_reflectance(sunz, satz, ssadiff, 0.45, redband)
```
```suggestion
  >>> refl_cor_m2 = virrs.get_reflectance(sunz, satz, ssadiff, 0.45, redband)
```

mraspaud commented on doc/rayleigh_correction.rst:
```diff
@@ -110,9 +109,32 @@ if you want another setup, e.g.:
+  >>> refl_cor_m2 = ufo.get_reflectance(sunz, satz, ssadiff, 0.45, redband)
+  [[ 10.40291763 9.654881]
+   [ 30.9275331  39.41288558]]
+  >>> reduced_refl_cor_m2 = ufo.reduce_rayleigh_highzenith(sunz, refl_cor_m2, 70., 90., 1.)
```
```suggestion
  >>> reduced_refl_cor_m2 = viirs.reduce_rayleigh_highzenith(sunz, refl_cor_m2, 70., 90., 1.)
```

mraspaud commented on pyspectral/rayleigh.py:
```diff
@@ -205,7 +207,7 @@ def get_effective_wavelength(self, bandname):
             bandname = get_bandname_from_wavelength(self.sensor, bandname, rsr.rsr)
 
         wvl, resp = rsr.rsr[bandname][
-            'det-1']['wavelength'], rsr.rsr[bandname]['det-1']['response']
+                        'det-1']['wavelength'], rsr.rsr[bandname]['det-1']['response']
```
Might be nicer with
```python
        wvl, resp = (rsr.rsr[bandname]['det-1']['wavelength'], 
                     rsr.rsr[bandname]['det-1']['response'])
```

mraspaud reviewed (changes requested):
Nice job, some minor issues.

Also, I see that right now, the high zenith adjustment is optional but wouldn't be logical to have it activated by default? Like the adjustment that is done in the edge of the disk...

simonrp84 commented:
While trying to fix the mess with new pushes to `main` this somehow got closed. Re-opening with the same changes to the code, plus fixes for two of Martin's suggestinos above.

mraspaud reviewed (approved):
LGTM

simonrp84 commented:
Thanks all!

mraspaud commented:
Thank you for the PR!

yukaribbba commented:
How can I use this function in my satpy script?

yukaribbba commented:
@simonrp84 Sorry but are you still there?
